nextcord/application_command.py

```diff
@@ -1724,245 +1389,477 @@ def check_against_raw_payload(
-        if not isinstance(self.type, ApplicationCommandType):
-            raise ValueError(
-                f"{self.error_name} Command type is not set to a valid type."
-            )
-        if (
-            self.type in (ApplicationCommandType.user, ApplicationCommandType.message)
-            and self.children
-        ):
-            raise ValueError(
-                f"{self.error_name} This type of command cannot have children."
-            )
-        if self.description and self.type in (
-            ApplicationCommandType.user,
-            ApplicationCommandType.message,
-        ):
-            raise ValueError(
-                f"{self.error_name} This type of command cannot have a description."
-            )
-
-    # Methods that can end up running the callback.
 
-    async def call_autocomplete_from_interaction(self, interaction: Interaction):
-        if not self._state:
-            raise NotImplementedError("State hasn't been set, this isn't handled yet!")
-        await self.call_autocomplete(
-            self._state, interaction, interaction.data.get("options", {})
-        )
+    def from_callback(
+            self,
+            callback: Optional[Callable] = None,
+            option_class: Optional[Type[BaseCommandOption]] = BaseCommandOption
+    ) -> None:
+        super().from_callback(callback=callback, option_class=option_class)
 
     async def call_from_interaction(self, interaction: Interaction) -> None:
-        if not self._state:
-            raise NotImplementedError("State hasn't been set, this isn't handled yet!")
-        await self.call(self._state, interaction, interaction.data.get("options", {}))
-
-    async def call(
-        self,
-        state: ConnectionState,
-        interaction: Interaction,
-        option_data: List[Dict[str, Any]],
-    ) -> None:
-        try:
-            await super().call(state, interaction, option_data)
-            return
-        except InvalidCommandType:
-            if self.type is ApplicationCommandType.message:
-                callback = self.call_invoke_message
-            elif self.type is ApplicationCommandType.user:
-                callback = self.call_invoke_user
-            else:
-                raise InvalidCommandType(
-                    f"{self.type} is not a handled Application Command type."
-                )
+        """|coro|
+        Calls the callback via the given :class:`Interaction`, relying on the locally
+        stored :class:`ConnectionState` object.
 
-        await self._call_with_hooks(state, interaction, self, callback, (interaction,))
+        Parameters
+        ----------
+        interaction: :class:`Interaction`
+            Interaction corresponding to the use of the command.
+        """
+        await self.call(self._state, interaction)
 
-    async def call_invoke_message(self, interaction: Interaction) -> None:
+    async def call(self, state: ConnectionState, interaction: Interaction) -> None:
```

Comment:
could this and all subclasses implement `__call__`? so the callback can be called manually like its a normal function if for ever the user needs that, similar to how it is for ext.commands

it would most likely need to take `Client` or get state off interaction

Reply:
Alrighty then. It should get state when the cog/client rolls through and does the from_callback/registration stuff though.